bench/BenchAll.hs

```diff
@@ -79,14 +87,57 @@ smallIntegerData = map fromIntegral intData
 largeIntegerData :: [Integer]
 largeIntegerData = map (* (10 ^ (100 :: Integer))) smallIntegerData
 
+{-# NOINLINE floatPosData #-}
+floatPosData :: [Float]
+floatPosData = map evenlyDistribute intData
+  where
+  evenlyDistribute :: Int -> Float
+  evenlyDistribute x = castWord32ToFloat $ increment * fromIntegral x
+  increment = castFloatToWord32 maxFinite `div` fromIntegral nRepl
 
-{-# NOINLINE floatData #-}
-floatData :: [Float]
-floatData = map (\x -> (3.14159 * fromIntegral x) ^ (3 :: Int)) intData
+{-# NOINLINE floatNegData #-}
+floatNegData :: [Float]
+floatNegData = map negate floatPosData
 
-{-# NOINLINE doubleData #-}
-doubleData :: [Double]
-doubleData = map (\x -> (3.14159 * fromIntegral x) ^ (3 :: Int)) intData
+{-# NOINLINE floatSpecials #-}
+floatSpecials :: [Float]
+floatSpecials = foldMap (const specials) [1..nRepl `div` length specials]
+  where
+  specials = [nan, infinity, negate infinity, 0 -0]
+
+{-# NOINLINE doublePosData #-}
+doublePosData :: [Double]
+doublePosData = map evenlyDistribute intData
+  where
+  evenlyDistribute :: Int -> Double
+  evenlyDistribute x = castWord64ToDouble $ increment * fromIntegral x
+  increment = (maximum - minimum) `div` fromIntegral nRepl
+  minimum = castDoubleToWord64 $ succIEEE $ 2 ^ 53
+  maximum = castDoubleToWord64 maxFinite 
+
+{-# NOINLINE doubleNegData #-}
+doubleNegData :: [Double]
+doubleNegData = map negate doublePosData
+
+-- f is an integer in the range [1, 2^53).
+{-# NOINLINE doublePosSmallData #-}
+doublePosSmallData :: [Double]
+doublePosSmallData = map evenlyDistribute intData
+  where
+  evenlyDistribute = assert (increment > 0) $ \x -> castWord64ToDouble $ increment * fromIntegral x + minimum
+  increment = (maximum - minimum) `div` fromIntegral nRepl
```

Comment:
This is more plausible. Factor it out and share the evenly-distributing with the other variants.

Reply:
Do you have any ideas for logarithmic incrementing? The large numbers dominate the range with a constant increment. I have split it up into Small and Big but I do not think this ideal.

Reply:
To the extent we need to care much about the composition of the benchmark inputs, splitting them up gives a good opportunity to demonstrate that, via a textual explanation of the split and/or via observably distinct run-times on different sides of the split.

But I'm not sure the current split is very precisely motivated; the small-ish numbers that we render differently are:
* The ones with decimal exponents in [-1, 6], for which we don't use scientific notation
* For Doubles only: The exact integers in [1, 2^53)